{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SGPS BioStatistics\n",
    "## Excercise 2: Getting publically availible data\n",
    "\n",
    "#### Gene Expression data from Gene Expression Omnibus\n",
    "+ https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE123336\n",
    "\n",
    "#### Specifically from this paper\n",
    "1.\tLaRocca D, Barns S, Hicks SD, Brindle A et al. Comparison of serum and saliva miRNAs for identification and characterization of mTBI in adult mixed martial arts fighters. *PLoS One* 2019;14(1):e0207785.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import the packages we need\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "# import GEOparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the data in\n",
    "human_expression = pd.read_csv('GSE123336_MMA_CountMatrix.csv.gz', index_col=0, compression='gzip')\n",
    "human_expression = human_expression.apply(pd.to_numeric, errors='coerce')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "22-Aug-2025 15:20:44 DEBUG utils - Directory . already exists. Skipping.\n",
      "22-Aug-2025 15:20:44 INFO GEOparse - Downloading ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE123nnn/GSE123336/soft/GSE123336_family.soft.gz to ./GSE123336_family.soft.gz\n",
      "100%|██████████████████████████████████████████████████████████████████████████████████████████████████| 12.6k/12.6k [00:00<00:00, 119kB/s]\n",
      "22-Aug-2025 15:20:44 DEBUG downloader - Size validation passed\n",
      "22-Aug-2025 15:20:44 DEBUG downloader - Moving /var/folders/j6/6pz4l2t10238hsq7sl5n5qww0000gn/T/tmpwlbl8jid to /Users/jpw/My Drive/teaching/2025_Biostatistics/e2_Getting_Data/GSE123336_family.soft.gz\n",
      "22-Aug-2025 15:20:44 DEBUG downloader - Successfully downloaded ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE123nnn/GSE123336/soft/GSE123336_family.soft.gz\n",
      "22-Aug-2025 15:20:44 INFO GEOparse - Parsing ./GSE123336_family.soft.gz: \n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - DATABASE: GeoMiame\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SERIES: GSE123336\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - PLATFORM: GPL18573\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500956\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500957\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500958\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500959\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500960\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500961\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500962\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500963\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500964\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500965\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500966\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500967\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500968\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500969\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500970\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500971\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500972\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500973\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500974\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500975\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500976\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500977\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500978\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500979\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500980\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500981\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500982\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500983\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500984\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500985\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500986\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500987\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500988\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500989\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500990\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500991\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500992\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500993\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500994\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500995\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500996\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500997\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500998\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3500999\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501000\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501001\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501002\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501003\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501004\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501005\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501006\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501007\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501008\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501009\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501010\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501011\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501012\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501013\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501014\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501015\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501016\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501017\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501018\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501019\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501020\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501021\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501022\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501023\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501024\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501025\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501026\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501027\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501028\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501029\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501030\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501031\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501032\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501033\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501034\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501035\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501036\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501037\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501038\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501039\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501040\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501041\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501042\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501043\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501044\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501045\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501046\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501047\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501048\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501049\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501050\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501051\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501052\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501053\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501054\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501055\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501056\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501057\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501058\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501059\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501060\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501061\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501062\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501063\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501064\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501065\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501066\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501067\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501068\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501069\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501070\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501071\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501072\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501073\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501074\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501075\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501076\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501077\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501078\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501079\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501080\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501081\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501082\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501083\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501084\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501085\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501086\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501087\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501088\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501089\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501090\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501091\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501092\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501093\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501094\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501095\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501096\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501097\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501098\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501099\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501100\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501101\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501102\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501103\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501104\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501105\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501106\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501107\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501108\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501109\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501110\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501111\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501112\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501113\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501114\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501115\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501116\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501117\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501118\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501119\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501120\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501121\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501122\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501123\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501124\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501125\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501126\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501127\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501128\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501129\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501130\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501131\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501132\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501133\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501134\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501135\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501136\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501137\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501138\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501139\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501140\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501141\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501142\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501143\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501144\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501145\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501146\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501147\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501148\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501149\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501150\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501151\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501152\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501153\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501154\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501155\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501156\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501157\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501158\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501159\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501160\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501161\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501162\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501163\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501164\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501165\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501166\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501167\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501168\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501169\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501170\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501171\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501172\n",
      "22-Aug-2025 15:20:44 DEBUG GEOparse - SAMPLE: GSM3501173\n"
     ]
    }
   ],
   "source": [
    "# Get the data straight (off the omnibus)\n",
    "gse = GEOparse.get_GEO(geo=\"GSE123336\", destdir=\".\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'GSM3500956': <SAMPLE: GSM3500956>,\n",
       " 'GSM3500957': <SAMPLE: GSM3500957>,\n",
       " 'GSM3500958': <SAMPLE: GSM3500958>,\n",
       " 'GSM3500959': <SAMPLE: GSM3500959>,\n",
       " 'GSM3500960': <SAMPLE: GSM3500960>,\n",
       " 'GSM3500961': <SAMPLE: GSM3500961>,\n",
       " 'GSM3500962': <SAMPLE: GSM3500962>,\n",
       " 'GSM3500963': <SAMPLE: GSM3500963>,\n",
       " 'GSM3500964': <SAMPLE: GSM3500964>,\n",
       " 'GSM3500965': <SAMPLE: GSM3500965>,\n",
       " 'GSM3500966': <SAMPLE: GSM3500966>,\n",
       " 'GSM3500967': <SAMPLE: GSM3500967>,\n",
       " 'GSM3500968': <SAMPLE: GSM3500968>,\n",
       " 'GSM3500969': <SAMPLE: GSM3500969>,\n",
       " 'GSM3500970': <SAMPLE: GSM3500970>,\n",
       " 'GSM3500971': <SAMPLE: GSM3500971>,\n",
       " 'GSM3500972': <SAMPLE: GSM3500972>,\n",
       " 'GSM3500973': <SAMPLE: GSM3500973>,\n",
       " 'GSM3500974': <SAMPLE: GSM3500974>,\n",
       " 'GSM3500975': <SAMPLE: GSM3500975>,\n",
       " 'GSM3500976': <SAMPLE: GSM3500976>,\n",
       " 'GSM3500977': <SAMPLE: GSM3500977>,\n",
       " 'GSM3500978': <SAMPLE: GSM3500978>,\n",
       " 'GSM3500979': <SAMPLE: GSM3500979>,\n",
       " 'GSM3500980': <SAMPLE: GSM3500980>,\n",
       " 'GSM3500981': <SAMPLE: GSM3500981>,\n",
       " 'GSM3500982': <SAMPLE: GSM3500982>,\n",
       " 'GSM3500983': <SAMPLE: GSM3500983>,\n",
       " 'GSM3500984': <SAMPLE: GSM3500984>,\n",
       " 'GSM3500985': <SAMPLE: GSM3500985>,\n",
       " 'GSM3500986': <SAMPLE: GSM3500986>,\n",
       " 'GSM3500987': <SAMPLE: GSM3500987>,\n",
       " 'GSM3500988': <SAMPLE: GSM3500988>,\n",
       " 'GSM3500989': <SAMPLE: GSM3500989>,\n",
       " 'GSM3500990': <SAMPLE: GSM3500990>,\n",
       " 'GSM3500991': <SAMPLE: GSM3500991>,\n",
       " 'GSM3500992': <SAMPLE: GSM3500992>,\n",
       " 'GSM3500993': <SAMPLE: GSM3500993>,\n",
       " 'GSM3500994': <SAMPLE: GSM3500994>,\n",
       " 'GSM3500995': <SAMPLE: GSM3500995>,\n",
       " 'GSM3500996': <SAMPLE: GSM3500996>,\n",
       " 'GSM3500997': <SAMPLE: GSM3500997>,\n",
       " 'GSM3500998': <SAMPLE: GSM3500998>,\n",
       " 'GSM3500999': <SAMPLE: GSM3500999>,\n",
       " 'GSM3501000': <SAMPLE: GSM3501000>,\n",
       " 'GSM3501001': <SAMPLE: GSM3501001>,\n",
       " 'GSM3501002': <SAMPLE: GSM3501002>,\n",
       " 'GSM3501003': <SAMPLE: GSM3501003>,\n",
       " 'GSM3501004': <SAMPLE: GSM3501004>,\n",
       " 'GSM3501005': <SAMPLE: GSM3501005>,\n",
       " 'GSM3501006': <SAMPLE: GSM3501006>,\n",
       " 'GSM3501007': <SAMPLE: GSM3501007>,\n",
       " 'GSM3501008': <SAMPLE: GSM3501008>,\n",
       " 'GSM3501009': <SAMPLE: GSM3501009>,\n",
       " 'GSM3501010': <SAMPLE: GSM3501010>,\n",
       " 'GSM3501011': <SAMPLE: GSM3501011>,\n",
       " 'GSM3501012': <SAMPLE: GSM3501012>,\n",
       " 'GSM3501013': <SAMPLE: GSM3501013>,\n",
       " 'GSM3501014': <SAMPLE: GSM3501014>,\n",
       " 'GSM3501015': <SAMPLE: GSM3501015>,\n",
       " 'GSM3501016': <SAMPLE: GSM3501016>,\n",
       " 'GSM3501017': <SAMPLE: GSM3501017>,\n",
       " 'GSM3501018': <SAMPLE: GSM3501018>,\n",
       " 'GSM3501019': <SAMPLE: GSM3501019>,\n",
       " 'GSM3501020': <SAMPLE: GSM3501020>,\n",
       " 'GSM3501021': <SAMPLE: GSM3501021>,\n",
       " 'GSM3501022': <SAMPLE: GSM3501022>,\n",
       " 'GSM3501023': <SAMPLE: GSM3501023>,\n",
       " 'GSM3501024': <SAMPLE: GSM3501024>,\n",
       " 'GSM3501025': <SAMPLE: GSM3501025>,\n",
       " 'GSM3501026': <SAMPLE: GSM3501026>,\n",
       " 'GSM3501027': <SAMPLE: GSM3501027>,\n",
       " 'GSM3501028': <SAMPLE: GSM3501028>,\n",
       " 'GSM3501029': <SAMPLE: GSM3501029>,\n",
       " 'GSM3501030': <SAMPLE: GSM3501030>,\n",
       " 'GSM3501031': <SAMPLE: GSM3501031>,\n",
       " 'GSM3501032': <SAMPLE: GSM3501032>,\n",
       " 'GSM3501033': <SAMPLE: GSM3501033>,\n",
       " 'GSM3501034': <SAMPLE: GSM3501034>,\n",
       " 'GSM3501035': <SAMPLE: GSM3501035>,\n",
       " 'GSM3501036': <SAMPLE: GSM3501036>,\n",
       " 'GSM3501037': <SAMPLE: GSM3501037>,\n",
       " 'GSM3501038': <SAMPLE: GSM3501038>,\n",
       " 'GSM3501039': <SAMPLE: GSM3501039>,\n",
       " 'GSM3501040': <SAMPLE: GSM3501040>,\n",
       " 'GSM3501041': <SAMPLE: GSM3501041>,\n",
       " 'GSM3501042': <SAMPLE: GSM3501042>,\n",
       " 'GSM3501043': <SAMPLE: GSM3501043>,\n",
       " 'GSM3501044': <SAMPLE: GSM3501044>,\n",
       " 'GSM3501045': <SAMPLE: GSM3501045>,\n",
       " 'GSM3501046': <SAMPLE: GSM3501046>,\n",
       " 'GSM3501047': <SAMPLE: GSM3501047>,\n",
       " 'GSM3501048': <SAMPLE: GSM3501048>,\n",
       " 'GSM3501049': <SAMPLE: GSM3501049>,\n",
       " 'GSM3501050': <SAMPLE: GSM3501050>,\n",
       " 'GSM3501051': <SAMPLE: GSM3501051>,\n",
       " 'GSM3501052': <SAMPLE: GSM3501052>,\n",
       " 'GSM3501053': <SAMPLE: GSM3501053>,\n",
       " 'GSM3501054': <SAMPLE: GSM3501054>,\n",
       " 'GSM3501055': <SAMPLE: GSM3501055>,\n",
       " 'GSM3501056': <SAMPLE: GSM3501056>,\n",
       " 'GSM3501057': <SAMPLE: GSM3501057>,\n",
       " 'GSM3501058': <SAMPLE: GSM3501058>,\n",
       " 'GSM3501059': <SAMPLE: GSM3501059>,\n",
       " 'GSM3501060': <SAMPLE: GSM3501060>,\n",
       " 'GSM3501061': <SAMPLE: GSM3501061>,\n",
       " 'GSM3501062': <SAMPLE: GSM3501062>,\n",
       " 'GSM3501063': <SAMPLE: GSM3501063>,\n",
       " 'GSM3501064': <SAMPLE: GSM3501064>,\n",
       " 'GSM3501065': <SAMPLE: GSM3501065>,\n",
       " 'GSM3501066': <SAMPLE: GSM3501066>,\n",
       " 'GSM3501067': <SAMPLE: GSM3501067>,\n",
       " 'GSM3501068': <SAMPLE: GSM3501068>,\n",
       " 'GSM3501069': <SAMPLE: GSM3501069>,\n",
       " 'GSM3501070': <SAMPLE: GSM3501070>,\n",
       " 'GSM3501071': <SAMPLE: GSM3501071>,\n",
       " 'GSM3501072': <SAMPLE: GSM3501072>,\n",
       " 'GSM3501073': <SAMPLE: GSM3501073>,\n",
       " 'GSM3501074': <SAMPLE: GSM3501074>,\n",
       " 'GSM3501075': <SAMPLE: GSM3501075>,\n",
       " 'GSM3501076': <SAMPLE: GSM3501076>,\n",
       " 'GSM3501077': <SAMPLE: GSM3501077>,\n",
       " 'GSM3501078': <SAMPLE: GSM3501078>,\n",
       " 'GSM3501079': <SAMPLE: GSM3501079>,\n",
       " 'GSM3501080': <SAMPLE: GSM3501080>,\n",
       " 'GSM3501081': <SAMPLE: GSM3501081>,\n",
       " 'GSM3501082': <SAMPLE: GSM3501082>,\n",
       " 'GSM3501083': <SAMPLE: GSM3501083>,\n",
       " 'GSM3501084': <SAMPLE: GSM3501084>,\n",
       " 'GSM3501085': <SAMPLE: GSM3501085>,\n",
       " 'GSM3501086': <SAMPLE: GSM3501086>,\n",
       " 'GSM3501087': <SAMPLE: GSM3501087>,\n",
       " 'GSM3501088': <SAMPLE: GSM3501088>,\n",
       " 'GSM3501089': <SAMPLE: GSM3501089>,\n",
       " 'GSM3501090': <SAMPLE: GSM3501090>,\n",
       " 'GSM3501091': <SAMPLE: GSM3501091>,\n",
       " 'GSM3501092': <SAMPLE: GSM3501092>,\n",
       " 'GSM3501093': <SAMPLE: GSM3501093>,\n",
       " 'GSM3501094': <SAMPLE: GSM3501094>,\n",
       " 'GSM3501095': <SAMPLE: GSM3501095>,\n",
       " 'GSM3501096': <SAMPLE: GSM3501096>,\n",
       " 'GSM3501097': <SAMPLE: GSM3501097>,\n",
       " 'GSM3501098': <SAMPLE: GSM3501098>,\n",
       " 'GSM3501099': <SAMPLE: GSM3501099>,\n",
       " 'GSM3501100': <SAMPLE: GSM3501100>,\n",
       " 'GSM3501101': <SAMPLE: GSM3501101>,\n",
       " 'GSM3501102': <SAMPLE: GSM3501102>,\n",
       " 'GSM3501103': <SAMPLE: GSM3501103>,\n",
       " 'GSM3501104': <SAMPLE: GSM3501104>,\n",
       " 'GSM3501105': <SAMPLE: GSM3501105>,\n",
       " 'GSM3501106': <SAMPLE: GSM3501106>,\n",
       " 'GSM3501107': <SAMPLE: GSM3501107>,\n",
       " 'GSM3501108': <SAMPLE: GSM3501108>,\n",
       " 'GSM3501109': <SAMPLE: GSM3501109>,\n",
       " 'GSM3501110': <SAMPLE: GSM3501110>,\n",
       " 'GSM3501111': <SAMPLE: GSM3501111>,\n",
       " 'GSM3501112': <SAMPLE: GSM3501112>,\n",
       " 'GSM3501113': <SAMPLE: GSM3501113>,\n",
       " 'GSM3501114': <SAMPLE: GSM3501114>,\n",
       " 'GSM3501115': <SAMPLE: GSM3501115>,\n",
       " 'GSM3501116': <SAMPLE: GSM3501116>,\n",
       " 'GSM3501117': <SAMPLE: GSM3501117>,\n",
       " 'GSM3501118': <SAMPLE: GSM3501118>,\n",
       " 'GSM3501119': <SAMPLE: GSM3501119>,\n",
       " 'GSM3501120': <SAMPLE: GSM3501120>,\n",
       " 'GSM3501121': <SAMPLE: GSM3501121>,\n",
       " 'GSM3501122': <SAMPLE: GSM3501122>,\n",
       " 'GSM3501123': <SAMPLE: GSM3501123>,\n",
       " 'GSM3501124': <SAMPLE: GSM3501124>,\n",
       " 'GSM3501125': <SAMPLE: GSM3501125>,\n",
       " 'GSM3501126': <SAMPLE: GSM3501126>,\n",
       " 'GSM3501127': <SAMPLE: GSM3501127>,\n",
       " 'GSM3501128': <SAMPLE: GSM3501128>,\n",
       " 'GSM3501129': <SAMPLE: GSM3501129>,\n",
       " 'GSM3501130': <SAMPLE: GSM3501130>,\n",
       " 'GSM3501131': <SAMPLE: GSM3501131>,\n",
       " 'GSM3501132': <SAMPLE: GSM3501132>,\n",
       " 'GSM3501133': <SAMPLE: GSM3501133>,\n",
       " 'GSM3501134': <SAMPLE: GSM3501134>,\n",
       " 'GSM3501135': <SAMPLE: GSM3501135>,\n",
       " 'GSM3501136': <SAMPLE: GSM3501136>,\n",
       " 'GSM3501137': <SAMPLE: GSM3501137>,\n",
       " 'GSM3501138': <SAMPLE: GSM3501138>,\n",
       " 'GSM3501139': <SAMPLE: GSM3501139>,\n",
       " 'GSM3501140': <SAMPLE: GSM3501140>,\n",
       " 'GSM3501141': <SAMPLE: GSM3501141>,\n",
       " 'GSM3501142': <SAMPLE: GSM3501142>,\n",
       " 'GSM3501143': <SAMPLE: GSM3501143>,\n",
       " 'GSM3501144': <SAMPLE: GSM3501144>,\n",
       " 'GSM3501145': <SAMPLE: GSM3501145>,\n",
       " 'GSM3501146': <SAMPLE: GSM3501146>,\n",
       " 'GSM3501147': <SAMPLE: GSM3501147>,\n",
       " 'GSM3501148': <SAMPLE: GSM3501148>,\n",
       " 'GSM3501149': <SAMPLE: GSM3501149>,\n",
       " 'GSM3501150': <SAMPLE: GSM3501150>,\n",
       " 'GSM3501151': <SAMPLE: GSM3501151>,\n",
       " 'GSM3501152': <SAMPLE: GSM3501152>,\n",
       " 'GSM3501153': <SAMPLE: GSM3501153>,\n",
       " 'GSM3501154': <SAMPLE: GSM3501154>,\n",
       " 'GSM3501155': <SAMPLE: GSM3501155>,\n",
       " 'GSM3501156': <SAMPLE: GSM3501156>,\n",
       " 'GSM3501157': <SAMPLE: GSM3501157>,\n",
       " 'GSM3501158': <SAMPLE: GSM3501158>,\n",
       " 'GSM3501159': <SAMPLE: GSM3501159>,\n",
       " 'GSM3501160': <SAMPLE: GSM3501160>,\n",
       " 'GSM3501161': <SAMPLE: GSM3501161>,\n",
       " 'GSM3501162': <SAMPLE: GSM3501162>,\n",
       " 'GSM3501163': <SAMPLE: GSM3501163>,\n",
       " 'GSM3501164': <SAMPLE: GSM3501164>,\n",
       " 'GSM3501165': <SAMPLE: GSM3501165>,\n",
       " 'GSM3501166': <SAMPLE: GSM3501166>,\n",
       " 'GSM3501167': <SAMPLE: GSM3501167>,\n",
       " 'GSM3501168': <SAMPLE: GSM3501168>,\n",
       " 'GSM3501169': <SAMPLE: GSM3501169>,\n",
       " 'GSM3501170': <SAMPLE: GSM3501170>,\n",
       " 'GSM3501171': <SAMPLE: GSM3501171>,\n",
       " 'GSM3501172': <SAMPLE: GSM3501172>,\n",
       " 'GSM3501173': <SAMPLE: GSM3501173>}"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gse.gsms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'title': ['MMA001-2533'],\n",
       " 'geo_accession': ['GSM3500956'],\n",
       " 'status': ['Public on Dec 05 2018'],\n",
       " 'submission_date': ['Dec 04 2018'],\n",
       " 'last_update_date': ['Dec 05 2018'],\n",
       " 'type': ['SRA'],\n",
       " 'channel_count': ['1'],\n",
       " 'source_name_ch1': ['Saliva'],\n",
       " 'organism_ch1': ['Homo sapiens'],\n",
       " 'taxid_ch1': ['9606'],\n",
       " 'characteristics_ch1': ['tissue: Saliva',\n",
       "  'timepoint: 0d post',\n",
       "  'hits to the head: 2',\n",
       "  'subject: MMA001'],\n",
       " 'molecule_ch1': ['total RNA'],\n",
       " 'extract_protocol_ch1': ['Saliva was collected in an RNA stabilizing container (Oragene RNA RE-100, DNA Genotek) and RNA was harvested using Qiazol reagent followed by RNeasy (Qiagen).',\n",
       "  'llumina TruSeq RNA Sample Prep Kit (Cat#FC-122-1001) was used with 250-500 nanograms of total RNA for the construction of sequencing libraries.'],\n",
       " 'description': ['Sample 1'],\n",
       " 'data_processing': ['Raw sequencing intensities were obtained on a NextSeq500 instrument and converted to base calls using bcl2fastq software (Illumina)',\n",
       "  'TruSeq Small RNA adapters were trimmed using cutadapt in Partek Flow (Partek), with minimum length set to 19',\n",
       "  'Alignment to miRbase 21 index of hg38 was performed using Shrimp2 in Partek Flow (Partek)',\n",
       "  'Quantification was performed in Partek Flow (Partek)',\n",
       "  'Normalization was performed to total read counts',\n",
       "  'Genome_build: hg38',\n",
       "  'Supplementary_files_format_and_content: a single data matrix of raw count abundance is provided as a comma separated value (CSV) file for the combined set of 218 samples'],\n",
       " 'platform_id': ['GPL18573'],\n",
       " 'contact_name': ['Frank,A,Middleton'],\n",
       " 'contact_email': ['middletf@upstate.edu'],\n",
       " 'contact_phone': ['315-464-7721'],\n",
       " 'contact_department': ['Neuroscience and Physiology'],\n",
       " 'contact_institute': ['SUNY Upstate Medical Univ'],\n",
       " 'contact_address': ['505 Irving Avenue'],\n",
       " 'contact_city': ['Syracuse'],\n",
       " 'contact_state': ['NY'],\n",
       " 'contact_zip/postal_code': ['13210'],\n",
       " 'contact_country': ['USA'],\n",
       " 'instrument_model': ['Illumina NextSeq 500'],\n",
       " 'library_selection': ['size fractionation'],\n",
       " 'library_source': ['transcriptomic'],\n",
       " 'library_strategy': ['miRNA-Seq'],\n",
       " 'relation': ['BioSample: https://www.ncbi.nlm.nih.gov/biosample/SAMN10520027',\n",
       "  'SRA: https://www.ncbi.nlm.nih.gov/sra?term=SRX5095235'],\n",
       " 'supplementary_file_1': ['NONE'],\n",
       " 'series_id': ['GSE123336'],\n",
       " 'data_row_count': ['0']}"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gse.gsms['GSM3500956'].metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# And collect the metadata\n",
    "meta = {}\n",
    "for key in gse.gsms:\n",
    "    # print(key)\n",
    "    samp = gse.gsms[key].metadata['description'][0]\n",
    "    characteristics = {}\n",
    "    for item in gse.gsms[key].metadata['characteristics_ch1']:\n",
    "        temp = item.split(': ')\n",
    "        characteristics[temp[0]] = temp[1]\n",
    "\n",
    "    if samp not in meta:\n",
    "        meta[samp] = {}\n",
    "        meta[samp] = characteristics\n",
    "    else:\n",
    "        meta[samp] = characteristics\n",
    "\n",
    "metadata = pd.DataFrame(meta).T\n",
    "metadata.to_csv('GSE123336_metadata.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tissue</th>\n",
       "      <th>timepoint</th>\n",
       "      <th>hits to the head</th>\n",
       "      <th>subject</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Sample 1</th>\n",
       "      <td>Saliva</td>\n",
       "      <td>0d post</td>\n",
       "      <td>2</td>\n",
       "      <td>MMA001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample 2</th>\n",
       "      <td>Serum</td>\n",
       "      <td>0d post</td>\n",
       "      <td>2</td>\n",
       "      <td>MMA001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample 3</th>\n",
       "      <td>Saliva</td>\n",
       "      <td>0d pre</td>\n",
       "      <td>0</td>\n",
       "      <td>MMA001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample 4</th>\n",
       "      <td>Serum</td>\n",
       "      <td>0d pre</td>\n",
       "      <td>0</td>\n",
       "      <td>MMA001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample 5</th>\n",
       "      <td>Saliva</td>\n",
       "      <td>1wk post</td>\n",
       "      <td>2</td>\n",
       "      <td>MMA001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample 214</th>\n",
       "      <td>Serum</td>\n",
       "      <td>0d pre</td>\n",
       "      <td>0</td>\n",
       "      <td>MMA040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample 215</th>\n",
       "      <td>Serum</td>\n",
       "      <td>0d post</td>\n",
       "      <td>35</td>\n",
       "      <td>MMA041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample 216</th>\n",
       "      <td>Serum</td>\n",
       "      <td>0d pre</td>\n",
       "      <td>0</td>\n",
       "      <td>MMA041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample 217</th>\n",
       "      <td>Serum</td>\n",
       "      <td>0d post</td>\n",
       "      <td>7</td>\n",
       "      <td>MMA042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample 218</th>\n",
       "      <td>Serum</td>\n",
       "      <td>0d pre</td>\n",
       "      <td>0</td>\n",
       "      <td>MMA042</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>218 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            tissue timepoint  hits to the head subject\n",
       "Sample 1    Saliva   0d post                 2  MMA001\n",
       "Sample 2     Serum   0d post                 2  MMA001\n",
       "Sample 3    Saliva    0d pre                 0  MMA001\n",
       "Sample 4     Serum    0d pre                 0  MMA001\n",
       "Sample 5    Saliva  1wk post                 2  MMA001\n",
       "...            ...       ...               ...     ...\n",
       "Sample 214   Serum    0d pre                 0  MMA040\n",
       "Sample 215   Serum   0d post                35  MMA041\n",
       "Sample 216   Serum    0d pre                 0  MMA041\n",
       "Sample 217   Serum   0d post                 7  MMA042\n",
       "Sample 218   Serum    0d pre                 0  MMA042\n",
       "\n",
       "[218 rows x 4 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata = pd.read_csv('GSE123336_metadata.csv', index_col=0)\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Can we see which miRNA are differentially expressed in Saliva pre and post fight?\n",
    "# Filter for saliva samples only\n",
    "saliva_samples = metadata[(metadata['tissue'] == 'Saliva') & \n",
    "                          (metadata['timepoint'].isin(['0d pre', '0d post']))]\n",
    "\n",
    "# Get the corresponding sample names\n",
    "saliva_sample_names = saliva_samples.index\n",
    "\n",
    "# Subset the expression data\n",
    "saliva_expression = human_expression[saliva_sample_names]\n",
    "saliva_expression\n",
    "\n",
    "pre_fight_samples = saliva_samples[saliva_samples['timepoint'] == '0d pre'].index\n",
    "post_fight_samples = saliva_samples[saliva_samples['timepoint'] == '0d post'].index\n",
    "\n",
    "pre_fight_expression = saliva_expression[pre_fight_samples]\n",
    "post_fight_expression = saliva_expression[post_fight_samples]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>miRNA</th>\n",
       "      <th>statistic</th>\n",
       "      <th>p_value</th>\n",
       "      <th>Significant</th>\n",
       "      <th>adjusted_p_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1287</th>\n",
       "      <td>hsa-miR-486-3p</td>\n",
       "      <td>96.0</td>\n",
       "      <td>0.001339</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>hsa-miR-204-5p</td>\n",
       "      <td>108.0</td>\n",
       "      <td>0.004007</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>960</th>\n",
       "      <td>hsa-miR-449c-5p</td>\n",
       "      <td>121.5</td>\n",
       "      <td>0.007006</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>992</th>\n",
       "      <td>hsa-miR-4524b-3p</td>\n",
       "      <td>132.0</td>\n",
       "      <td>0.010144</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>hsa-miR-1185-2-3p</td>\n",
       "      <td>142.0</td>\n",
       "      <td>0.011817</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1992</th>\n",
       "      <td>hsa-miR-6805-3p</td>\n",
       "      <td>208.5</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>547</th>\n",
       "      <td>hsa-miR-3155a</td>\n",
       "      <td>209.5</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>322</th>\n",
       "      <td>hsa-miR-196b-3p</td>\n",
       "      <td>209.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325</th>\n",
       "      <td>hsa-miR-197-5p</td>\n",
       "      <td>208.5</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1893</th>\n",
       "      <td>hsa-miR-6757-5p</td>\n",
       "      <td>208.5</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2347 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  miRNA  statistic   p_value  Significant  adjusted_p_value\n",
       "1287     hsa-miR-486-3p       96.0  0.001339        False               1.0\n",
       "351      hsa-miR-204-5p      108.0  0.004007        False               1.0\n",
       "960     hsa-miR-449c-5p      121.5  0.007006        False               1.0\n",
       "992    hsa-miR-4524b-3p      132.0  0.010144        False               1.0\n",
       "47    hsa-miR-1185-2-3p      142.0  0.011817        False               1.0\n",
       "...                 ...        ...       ...          ...               ...\n",
       "1992    hsa-miR-6805-3p      208.5  1.000000        False               1.0\n",
       "547       hsa-miR-3155a      209.5  1.000000        False               1.0\n",
       "322     hsa-miR-196b-3p      209.0  1.000000        False               1.0\n",
       "325      hsa-miR-197-5p      208.5  1.000000        False               1.0\n",
       "1893    hsa-miR-6757-5p      208.5  1.000000        False               1.0\n",
       "\n",
       "[2347 rows x 5 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from scipy.stats import mannwhitneyu\n",
    "\n",
    "# Initialize a list to store results\n",
    "diff_expression_results = []\n",
    "\n",
    "# Iterate through each miRNA\n",
    "for miRNA in saliva_expression.index:\n",
    "    # Extract expression levels for this miRNA\n",
    "    pre_values = pre_fight_expression.loc[miRNA]\n",
    "    post_values = post_fight_expression.loc[miRNA]\n",
    "    \n",
    "    # Perform the Mann-Whitney U test\n",
    "    stat, p_value = mannwhitneyu(pre_values, post_values, alternative='two-sided')\n",
    "    \n",
    "    # Store the results\n",
    "    diff_expression_results.append({\n",
    "        'miRNA': miRNA,\n",
    "        'statistic': stat,\n",
    "        'p_value': p_value\n",
    "    })\n",
    "\n",
    "# Convert the results to a DataFrame\n",
    "diff_expression_df = pd.DataFrame(diff_expression_results)\n",
    "\n",
    "# Apply multiple testing correction (e.g., Benjamini-Hochberg)\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "diff_expression_df['Significant'], diff_expression_df['adjusted_p_value'], _, _ = multipletests(diff_expression_df['p_value'], method='fdr_bh')\n",
    "\n",
    "# Sort the results by adjusted p-value\n",
    "diff_expression_df = diff_expression_df.sort_values('p_value')\n",
    "diff_expression_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Can we see which miRNA are differentially expressed in Serum pre and post fight?\n",
    "# Filter for serum samples only\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How about getting rat data for mTBI?\n",
    "#### Gene Expression data from Gene Expression Omnibus\n",
    "+ https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE159011\n",
    "\n",
    "#### Specifically from this paper\n",
    "1.\tDas Gupta S, Ciszek R, Heiskanen M, Lapinlampi N et al. Plasma miR-9-3p and miR-136-3p as Potential Novel Diagnostic Biomarkers for Experimental and Human Mild Traumatic Brain Injury. *Int J Mol Sci* 2021 Feb 4;22(4).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rat_expression = pd.read_csv(\"GSE159011_Raw_counts_matrix.txt\", index_col=0, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rat_expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rat_expression.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normalization\n",
    "# Step 1: Calculate the sum of each column\n",
    "column_sums = rat_expression.sum()\n",
    "\n",
    "# Step 2: Determine the normalization factor to scale each column to sum to 1 million\n",
    "normalization_factor = 1_000_000 / column_sums\n",
    "\n",
    "# Step 3: Multiply each column by its respective normalization factor\n",
    "normalized_df = rat_expression.multiply(normalization_factor, axis=1)\n",
    "\n",
    "# Now each column in `normalized_df` should sum to 1 million\n",
    "normalized_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We also want to do put this on a log2 scale\n",
    "log2_df = np.log2(normalized_df + 1)\n",
    "log2_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "22-Aug-2025 15:21:37 DEBUG utils - Directory . already exists. Skipping.\n",
      "22-Aug-2025 15:21:37 INFO GEOparse - Downloading ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE159nnn/GSE159011/soft/GSE159011_family.soft.gz to ./GSE159011_family.soft.gz\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████| 4.47k/4.47k [00:00<00:00, 39.7kB/s]\n",
      "22-Aug-2025 15:21:37 DEBUG downloader - Size validation passed\n",
      "22-Aug-2025 15:21:37 DEBUG downloader - Moving /var/folders/j6/6pz4l2t10238hsq7sl5n5qww0000gn/T/tmpkvqpjbab to /Users/jpw/My Drive/teaching/2025_Biostatistics/e2_Getting_Data/GSE159011_family.soft.gz\n",
      "22-Aug-2025 15:21:37 DEBUG downloader - Successfully downloaded ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE159nnn/GSE159011/soft/GSE159011_family.soft.gz\n",
      "22-Aug-2025 15:21:37 INFO GEOparse - Parsing ./GSE159011_family.soft.gz: \n",
      "22-Aug-2025 15:21:37 DEBUG GEOparse - DATABASE: GeoMiame\n",
      "22-Aug-2025 15:21:37 DEBUG GEOparse - SERIES: GSE159011\n",
      "22-Aug-2025 15:21:37 DEBUG GEOparse - PLATFORM: GPL25029\n",
      "22-Aug-2025 15:21:37 DEBUG GEOparse - SAMPLE: GSM4817312\n",
      "22-Aug-2025 15:21:37 DEBUG GEOparse - SAMPLE: GSM4817313\n",
      "22-Aug-2025 15:21:37 DEBUG GEOparse - SAMPLE: GSM4817314\n",
      "22-Aug-2025 15:21:37 DEBUG GEOparse - SAMPLE: GSM4817315\n",
      "22-Aug-2025 15:21:37 DEBUG GEOparse - SAMPLE: GSM4817316\n",
      "22-Aug-2025 15:21:37 DEBUG GEOparse - SAMPLE: GSM4817317\n",
      "22-Aug-2025 15:21:37 DEBUG GEOparse - SAMPLE: GSM4817318\n",
      "22-Aug-2025 15:21:37 DEBUG GEOparse - SAMPLE: GSM4817319\n",
      "22-Aug-2025 15:21:37 DEBUG GEOparse - SAMPLE: GSM4817320\n",
      "22-Aug-2025 15:21:37 DEBUG GEOparse - SAMPLE: GSM4817321\n",
      "22-Aug-2025 15:21:37 DEBUG GEOparse - SAMPLE: GSM4817322\n",
      "22-Aug-2025 15:21:37 DEBUG GEOparse - SAMPLE: GSM4817323\n",
      "22-Aug-2025 15:21:37 DEBUG GEOparse - SAMPLE: GSM4817324\n",
      "22-Aug-2025 15:21:37 DEBUG GEOparse - SAMPLE: GSM4817325\n",
      "22-Aug-2025 15:21:37 DEBUG GEOparse - SAMPLE: GSM4817326\n",
      "22-Aug-2025 15:21:37 DEBUG GEOparse - SAMPLE: GSM4817327\n",
      "22-Aug-2025 15:21:37 DEBUG GEOparse - SAMPLE: GSM4817328\n",
      "22-Aug-2025 15:21:37 DEBUG GEOparse - SAMPLE: GSM4817329\n",
      "22-Aug-2025 15:21:37 DEBUG GEOparse - SAMPLE: GSM4817330\n"
     ]
    }
   ],
   "source": [
    "# Get the data straight (off the omnibus)\n",
    "# gse = GEOparse.get_GEO(geo=\"GSE159011\", destdir=\".\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# And collect the metadata\n",
    "meta = {}\n",
    "for key in gse.gsms:\n",
    "    # print(key)\n",
    "    samp = gse.gsms[key].metadata['description'][0]\n",
    "    characteristics = {}\n",
    "    for item in gse.gsms[key].metadata['characteristics_ch1']:\n",
    "        temp = item.split(': ')\n",
    "        characteristics[temp[0]] = temp[1]\n",
    "\n",
    "    if samp not in meta:\n",
    "        meta[samp] = {}\n",
    "        meta[samp] = characteristics\n",
    "    else:\n",
    "        meta[samp] = characteristics\n",
    "\n",
    "rat_metadata = pd.DataFrame(meta).T\n",
    "rat_metadata.to_csv('GSE159011_metadata.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rat_metadata = pd.read_csv('GSE159011_metadata.csv', index_col=0)\n",
    "rat_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter metadata for Naive and Severe TBI groups\n",
    "log2_df.columns = log2_df.columns.str.replace('-UMIs', '')\n",
    "naive_samples = rat_metadata[rat_metadata['group'] == 'Naive'].index\n",
    "severe_tbi_samples = rat_metadata[rat_metadata['group'] == 'Severe TBI'].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset expression data for Naive and Severe TBI samples\n",
    "naive_expression = log2_df[naive_samples]\n",
    "severe_tbi_expression = log2_df[severe_tbi_samples]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize a list to store results\n",
    "diff_expression_results = []\n",
    "\n",
    "# Iterate through each miRNA\n",
    "for miRNA in log2_df.index:\n",
    "    # Extract expression levels for this miRNA in Naive and Severe TBI groups\n",
    "    naive_values = naive_expression.loc[miRNA]\n",
    "    severe_tbi_values = severe_tbi_expression.loc[miRNA]\n",
    "    \n",
    "    # Perform the Mann-Whitney U test\n",
    "    stat, p_value = mannwhitneyu(naive_values, severe_tbi_values, alternative='two-sided')\n",
    "    \n",
    "    # Store the results\n",
    "    diff_expression_results.append({\n",
    "        'miRNA': miRNA,\n",
    "        'statistic': stat,\n",
    "        'p_value': p_value\n",
    "    })\n",
    "\n",
    "# Convert the results to a DataFrame\n",
    "diff_expression_df = pd.DataFrame(diff_expression_results)\n",
    "\n",
    "# Drop out NaN from the p-values\n",
    "diff_expression_df = diff_expression_df.dropna(subset=['p_value'])\n",
    "\n",
    "# Apply multiple testing correction (e.g., Benjamini-Hochberg)\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "diff_expression_df['adjusted_p_value'] = multipletests(diff_expression_df['p_value'], method='fdr_bh')[1]\n",
    "\n",
    "# Apply multiple testing correction (e.g., Benjamini-Hochberg)\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "diff_expression_df['Significant'], diff_expression_df['adjusted_p_value'], _, _ = multipletests(diff_expression_df['p_value'], method='fdr_bh')\n",
    "\n",
    "# Sort the results by adjusted p-value\n",
    "diff_expression_df = diff_expression_df.sort_values('p_value')\n",
    "diff_expression_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
